presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -88,401 +110,731 @@
         this.session = requireNonNull(session, "session is null");
         this.sqlParser = requireNonNull(sqlParser, "sql parser is null");
         this.accessControl = requireNonNull(accessControl, "access control is null");
         this.domainTranslator = requireNonNull(domainTranslator, "row expression domain translator is null");
-        this.materializedView = requireNonNull(materializedView, "materialized view is null");
-        this.materializedViewQuery = requireNonNull(materializedViewQuery, "materialized view query is null");
+        this.baseTableToMaterializedViewMap = requireNonNull(baseTableToMaterializedViewMap, "base table to materialized view map is null");
         FunctionAndTypeManager functionAndTypeManager = metadata.getFunctionAndTypeManager();
         logicalRowExpressions = new LogicalRowExpressions(
                 new RowExpressionDeterminismEvaluator(functionAndTypeManager),
                 new FunctionResolution(functionAndTypeManager),
                 functionAndTypeManager);
     }
 
-    public Node rewrite(Node node)
+    public Query rewrite(Query node)
+    {
+        return (Query) process(node);
+    }
+
+    @Override
+    protected Node visitQuerySpecification(QuerySpecification node, Void context)
     {
-        // TODO: Implement ways to handle non-optimizable query without throw/catch. https://github.com/prestodb/presto/issues/16541
-        try {
-            MaterializedViewInformationExtractor materializedViewInformationExtractor = new MaterializedViewInformationExtractor();
-            materializedViewInformationExtractor.process(materializedViewQuery);
-            materializedViewInfo = materializedViewInformationExtractor.getMaterializedViewInfo();
-            return process(node);
-        }
-        catch (Exception ex) {
-            logger.warn("Failed to rewrite query with materialized view with following exception: %s", ex.getMessage());
+        if (!node.getFrom().isPresent()) {
+            return node;
+        }
+        // If a query specification has a Table as a source, it can potentially be rewritten,
+        // so hand control over to LeafSubqueryVisitor via rewriteQuerySpecificationIfCompatible
+        Table baseTable;
+        Relation from = node.getFrom().get();
+        if (from instanceof Table) {
+            baseTable = (Table) from;
+            return rewriteQuerySpecificationIfCompatible(node, baseTable);
+        }
+        if (from instanceof AliasedRelation
+                && ((AliasedRelation) from).getRelation() instanceof Table) {
+            AliasedRelation aliasedRelation = (AliasedRelation) from;
+            baseTable = (Table) aliasedRelation.getRelation();
+            return rewriteQuerySpecificationIfCompatible(node, baseTable);
+        }
+
+        if (!node.getFrom().isPresent()) {
+            return node;
+        }
+        Relation newFrom = processSameType(from);
```

Comment:
There should be a blank line before the line `Relation newFrom = processSameType(from);` instead of after. This line is more logically related to the next `if` .